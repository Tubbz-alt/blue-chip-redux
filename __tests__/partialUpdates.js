import {resourcesReducer} from "..";
import normalizedJsonProjectPayload from "../__testHelpers__/fixtures/normalizedJsonProjectPayload";

describe("Partial Update tests:", () => {
  describe("partial update relationships:", () => {
    let state = {index: {}};

    it("should return the initial state", () => {
      expect(resourcesReducer(undefined, {})).toEqual({index: {}});
    });
    it("should update the store given a UPDATE_RESOURCES action", () => {
      const action = {
        resourceType: "projects",
        resourcesById: normalizedJsonProjectPayload,
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("should only update the relationships given, belongsTo scenario", () => {
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            relationships: {
              freightVendor: {
                data: {
                  id: "1",
                  type: "vendors"
                }
              }
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("should only update the relationships given, hasMany scenario", () => {
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            relationships: {
              specs: {
                data: [
                  {
                    id: "1",
                    type: "specs"
                  }
                ]
              }
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("should only update the relationships given, no relationships sent scenario", () => {
      const relationships = undefined;
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            relationships
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      delete action.resourcesById["1"].relationships;
      expect("relationships" in action.resourcesById["1"]).toBeFalsy();
      expect(state).toMatchSnapshot();
    });

    it("should add new relationship/update others sent", () => {
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            relationships: {
              newThing: {
                data: {
                  id: "1",
                  type: "newThing"
                }
              },
              specs: {
                data: [
                  {
                    id: "2",
                    type: "specs"
                  }
                ]
              }
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("should allow null and empty when given", () => {
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            relationships: {
              newThing: {
                data: null
              },
              specs: {
                data: []
              }
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("adds relationships in when previously not existing", () => {
      const action = {
        resourceType: "testObjects",
        resourcesById: {
          "1": {
            id: "1",
            type: "testObjects",
            attributes: {
              testField1: "testValue"
            },
            relationships: {
              newThing: {
                data: {id: "1", type: "newThings"}
              }
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(
        {index: { testObjects: ["1"] }, testObjects: {"1": {id: "1", type: "testObjects"}}},
        action
      );
      expect(state).toMatchSnapshot();
    });
  });

  describe("partial update attributes:", () => {
    let state = {index: {}};

    it("should update the store given a UPDATE_RESOURCES action", () => {
      const action = {
        resourceType: "projects",
        resourcesById: normalizedJsonProjectPayload,
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("should update only the attributes given", () => {
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            attributes: {
              fundingDue: "Yearly"
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });

    it("should add new attributes when given, update others given", () => {
      const action = {
        resourceType: "projects",
        resourcesById: {
          ...normalizedJsonProjectPayload,
          "1": {
            ...normalizedJsonProjectPayload["1"],
            attributes: {
              fundingDue: "Annually",
              newAttribute: "someValue"
            }
          }
        },
        type: "UPDATE_RESOURCES",
        index: ["1"]
      };
      state = resourcesReducer(state, action);
      expect(state).toMatchSnapshot();
    });
  });
});
