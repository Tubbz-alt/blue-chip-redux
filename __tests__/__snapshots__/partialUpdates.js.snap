// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partial Update tests: partial update attributes: should add new attributes when given, update others given 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Annually",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "newAttribute": "someValue",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "2",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
            Object {
              "id": "2",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update attributes: should update only the attributes given 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Yearly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "2",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
            Object {
              "id": "2",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update attributes: should update the store given a UPDATE_RESOURCES action 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "2",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
            Object {
              "id": "2",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: adds relationships in when previously not existing 1`] = `
Object {
  "index": Object {
    "testObjects": Array [
      "1",
    ],
  },
  "testObjects": Object {
    "1": Object {
      "attributes": Object {
        "testField1": "testValue",
      },
      "id": "1",
      "relationships": Object {
        "newThing": Object {
          "data": Object {
            "id": "1",
            "type": "newThings",
          },
        },
      },
      "type": "testObjects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: should add new relationship/update others sent 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "1",
            "type": "vendors",
          },
        },
        "newThing": Object {
          "data": Object {
            "id": "1",
            "type": "newThing",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "2",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: should allow null and empty when given 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "1",
            "type": "vendors",
          },
        },
        "newThing": Object {
          "data": null,
        },
        "specs": Object {
          "data": Array [],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: should only update the relationships given, belongsTo scenario 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "1",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
            Object {
              "id": "2",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: should only update the relationships given, hasMany scenario 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "1",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: should only update the relationships given, no relationships sent scenario 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "1",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;

exports[`Partial Update tests: partial update relationships: should update the store given a UPDATE_RESOURCES action 1`] = `
Object {
  "index": Object {
    "projects": Array [
      "1",
    ],
  },
  "projects": Object {
    "1": Object {
      "attributes": Object {
        "clientId": 2,
        "createdAt": "2015-06-14T22:12:05.275Z",
        "currencyId": 1,
        "endDate": "2015-06-14T22:12:05.275Z",
        "entityId": 1,
        "freightVendorId": 2,
        "fundingDue": "Monthly",
        "fundingType": "Cashflow",
        "isStarred": null,
        "name": "Project 1",
        "nickname": "nickname-1",
        "officeId": 1,
        "propertyId": 1,
        "scopeConsulting": false,
        "scopeFfe": false,
        "scopeModelRoom": false,
        "scopeOse": false,
        "startDate": "2015-06-14T22:12:05.275Z",
        "status": "Proposed",
        "teamId": 1,
        "type": "Full Renovation",
        "updatedAt": "2015-06-14T22:12:05.275Z",
      },
      "id": "1",
      "relationships": Object {
        "freightVendor": Object {
          "data": Object {
            "id": "2",
            "type": "vendors",
          },
        },
        "specs": Object {
          "data": Array [
            Object {
              "id": "1",
              "type": "specs",
            },
            Object {
              "id": "2",
              "type": "specs",
            },
          ],
        },
      },
      "type": "projects",
    },
  },
}
`;
